processing/src/main/java/io/druid/query/ChainedExecutionQueryRunner.java

```diff
@@ -70,87 +74,115 @@ public ChainedExecutionQueryRunner(
+                        try {
+                          Sequence<T> result = queryRunner.run(threadSafeQueryPlus, responseContext);
+                          if (result == null) {
+                            throw new ISE("Got a null result! Segments are missing!");
+                          }
+
+                          List<T> retVal = result.toList();
+                          if (retVal == null) {
+                            throw new ISE("Got a null list of results! WTF?!");
                           }
 
-                          return exec.submit(
-                              new AbstractPrioritizedCallable<Iterable<T>>(priority)
-                              {
-                                @Override
-                                public Iterable<T> call()
-                                {
-                                  try {
-                                    Sequence<T> result = input.run(threadSafeQueryPlus, responseContext);
-                                    if (result == null) {
-                                      throw new ISE("Got a null result! Segments are missing!");
-                                    }
-
-                                    List<T> retVal = result.toList();
-                                    if (retVal == null) {
-                                      throw new ISE("Got a null list of results! WTF?!");
-                                    }
-
-                                    return retVal;
-                                  }
-                                  catch (QueryInterruptedException e) {
-                                    throw Throwables.propagate(e);
-                                  }
-                                  catch (Exception e) {
-                                    log.error(e, "Exception with one of the sequences!");
-                                    throw Throwables.propagate(e);
-                                  }
-                                }
-                              }
-                          );
+                          return retVal;
                         }
-                    )
-                )
+                        catch (QueryInterruptedException e) {
+                          throw Throwables.propagate(e);
+                        }
+                        catch (Exception e) {
+                          log.error(e, "Exception with one of the sequences!");
+                          throw Throwables.propagate(e);
+                        }
+                      }
+                    }
+                ).map(exec::submit)
             );
 
             queryWatcher.registerQuery(query, futures);
 
             try {
+              final DateTime deadline;
+              if (QueryContexts.hasTimeout(query)) {
+                deadline = DateTimes.nowUtc().plusMillis((int) QueryContexts.getTimeout(query));
+              } else {
+                deadline = DateTimes.utc(JodaUtils.MAX_INSTANT);
```

Comment:
If you want to do this, please change all other places using query timeout as well. I think this should be a separate issue though.